### PR TITLE
feat(rpc): add progressive trace range and checkpoint helpers

### PR DESCRIPTION
**Pull request by danielntmd**
### **User description**
This PR adds the shared building blocks for tracing a block incrementally.

Progressive Cache Behavior: 
- Cached transaction traces provide the state checkpoint for executing the remaining transactions, and extensions preserve previously published results. 
- Requests for initial reads replay the full block (retained logic).
- Continuation checkpoints are rebuilt from cached per-transaction state diffs when needed. This trades recomputation during resumption for lower cache memory usage, avoiding a duplicate cumulative state diff.

Note: State resumption, checkpoint class loading, and execution-error offsets are covered by RPC integration tests in #4080.

___

### **PR Type**
Enhancement

___

### **Description**
- Introduce progressive trace range helpers to support incremental block tracing

- Reconstruct VM starting state checkpoints from cached per-transaction state diffs

- Add `Complete` flag to `BlockTrace` and update coverage validation logic

- Update RPC v9 and v10 handlers to explicitly mark full block traces as complete

___

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>trace_adaptation_test.go</strong><dd><code>Set record as complete in trace adaptation tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-ceec5b9e8cf412b6bdebbd3ddd59d4def796e54fcdedf01c43b6dfe2252c6537">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>progressive_test.go</strong><dd><code>Add tests for progressive state resumption and trace combination</code></dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-71e05777ebb8d63e0ba7be2472b01cf8f4dbdfaccf48713e7c1c2de8c9e2575c">+262/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>result_test.go</strong><dd><code>Update VM result tests to verify completion and target coverage</code></dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-e07a163d664d6cb37c2118498f5ba50c6a520b0dd98627bef7e11404dcaf6e66">+6/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>5 files</summary><table>
<tr>
  <td><strong>progressive.go</strong><dd><code>Implement progressive trace range planning and state resumption</code></dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-52f2075d90259e060168a36b96768cebab619d03043303711d4af4b7b4ee090d">+236/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>result.go</strong><dd><code>Mark feeder-sourced block traces as complete</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-a63ecd08405077899a7e474b78c8c365aa727fbc953d1c3e6e2cdab3c1b9fbfe">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>trace.go</strong><dd><code>Add Complete flag and TransactionTarget validation to BlockTrace</code></dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-065189a5fc9aa8e67236e1464dc1321a58ab0dc129fa999f65628934f9420e93">+33/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>trace.go</strong><dd><code>Mark traced blocks as complete in RPC v10 handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-bc866579e372f57f11f467f936326fbceeed552cb885b9f6350ac734b1429f4a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>trace.go</strong><dd><code>Mark traced blocks as complete in RPC v9 handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/NethermindEth/juno/pull/4079/files#diff-75589d22fdf6a694fd19ef5ac3eddbaa5983c609b0fe41f12e71aa3c0be52482">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: rpc/tracecache/progressive.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+// A six-transaction block has two traces cached. A request arrives for transaction 4:
+//   - Work needed: transactions [2, 5).
+//   - Starting state: if the cached transactions changed a nonce from 6 to 7 to 8,
+//     the VM starts with nonce 8.
+//   - Result: the cache now covers transactions 0 through 4. A later request can
+//     extend it to include transaction 5.
+//
+// See [Cache] for the mutability contract for published traces.
+type Range struct {
+	Start, End uint64
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:36]: nit: please one param per line. 

Honestly, in this case makes a lot of sense and it is readable so feel free to ignore. It is mostly a nitpick about keeping the "status quo" style

File: rpc/tracecache/progressive.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+//     RPC handler. Nil requests execution through the block's end.
+//   - initialReads: when true, the entire block must be replayed from the beginning.
+//     The target must be nil or the last transaction.
+//     Request initial reads from the VM and pass true to [FromVM].
+func PlanRange(
+	cached *BlockTrace,
+	transactions []core.Transaction,
+	target *TransactionTarget,
+	initialReads bool,
+) (*Range, error) {
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:59]: Why can `Range` be returned by value?

File: rpc/tracecache/progressive.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	if target != nil && (target.Hash == nil || target.Index >= uint64(len(transactions)) ||
+		!transactions[target.Index].Hash().Equal(target.Hash)) {
+		return nil, ErrTargetNotFound
+	}
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:63]: another style guide we follow is separating the conditions from the `if` (or `for`) statements if they become too big. 

Suggestion:
```go
validTarget := <long cond>
if validTarget {
  // ...
}
```

File: rpc/tracecache/progressive.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+// Reconstructs the state after the cached prefix so that the VM can execute
+// the remaining transactions without replaying that prefix.
+//
+// Parameters:
+//   - parent: the state reader for the state immediately before the block.
+//   - classes: the state reader used to load classes declared in the cached
+//     prefix, typically the chain's head state.
+//   - blockNumber: the number of the block being traced.
+func (r *Range) ResumeState(
+	parent, classes core.StateReader,
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:98]: one parameter per line

File: rpc/tracecache/progressive.go (hunk 1 above)
Comment (danielntmd) [reply rpc/tracecache/progressive.go]: I'll make the change to maintain parity in style.

File: rpc/tracecache/progressive.go (hunk 2 above)
Comment (danielntmd) [reply rpc/tracecache/progressive.go]: Returning by value here makes more sense.

File: rpc/tracecache/progressive.go (hunk 3 above)
Comment (danielntmd) [reply rpc/tracecache/progressive.go]: Separated the conditions from the `if`.

File: rpc/tracecache/trace.go [hunk 5]
```diff
@@ -37,12 +41,41 @@ func (t *TransactionTrace) FeederTrace() *starknet.TransactionTrace {
...
+	}
+	return nil
+}
+
+// TransactionTarget identifies the transaction to trace.
+// A nil target requests a full block trace.
+type TransactionTarget struct {
+	Index uint64
+	Hash  *felt.Felt
+}
```
Comment (rodrodros) [inline rpc/tracecache/trace.go:81]: Hash should be a `*felt.Hash` right? Or more specifically, a `*felt.TransactionHash`? 

It should avoid the conversion in the follow up PR

File: rpc/tracecache/trace.go [hunk 6]
```diff
@@ -37,12 +41,41 @@ func (t *TransactionTrace) FeederTrace() *starknet.TransactionTrace {
...
 func (b *BlockTrace) Covers(initialReads bool) bool {
+	return b.CoversTarget(nil, initialReads)
+}
+
+// CoversTarget checks coverage through the requested transaction and requested initial reads.
+// A nil target requires complete block coverage.
+func (b *BlockTrace) CoversTarget(target *TransactionTarget, initialReads bool) bool {
+	if !b.Complete && (target == nil || target.Index >= uint64(len(b.Traces))) {
+		return false
+	}
```
Comment (rodrodros) [inline rpc/tracecache/trace.go:64]: Nit: 

I believe that in this case legibility is more important at the cost of some code duplication. A method name `CoversTarget` shouldn't accept a nil target. This is a tradeoff taken to have only one impl and opens the case for people start using `CoversTarget` wrongly (instead of using `Covers`).

I would propose to separate the implementations and make sure both `Covers` and `CoversTarget` are well unit tested and using `CoversTarget` with a nil target becomes a nil pointer error.

File: rpc/tracecache/trace.go [hunk 7]
```diff
@@ -37,12 +41,41 @@ func (t *TransactionTrace) FeederTrace() *starknet.TransactionTrace {
...
+// ValidateTarget checks that the target index identifies a trace with the requested hash.
+func (b *BlockTrace) ValidateTarget(target *TransactionTarget) error {
+	invalidTarget := target != nil && (target.Hash == nil ||
+		target.Index >= uint64(len(b.Traces)) ||
+		!b.Traces[target.Index].Hash.Equal(target.Hash))
+	if invalidTarget {
+		return ErrTargetNotFound
+	}
+	return nil
+}
```
Comment (rodrodros) [inline rpc/tracecache/trace.go:74]: `ValidateTarget` returns an error but at not point this error is being used. If this is intended, then it is better to update the signature to return a boolean and perhaps update the name to `ValidTarget` or `IsValidTarget`.

File: rpc/tracecache/progressive.go [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+//     RPC handler. Nil requests execution through the block's end.
+//   - initialReads: when true, the entire block must be replayed from the beginning.
+//     The target must be nil or the last transaction.
+//     Request initial reads from the VM and pass true to [FromVM].
+func PlanRange(
+	cached *BlockTrace,
+	transactions []core.Transaction,
+	target *TransactionTarget,
+	initialReads bool,
+) (Range, error) {
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:60]: I am a bit at odds with target being an optional argument but I see there is just one caller for each version and I am not sure it makes sense to complicate the API more. It is ok like this 👍

File: rpc/tracecache/progressive.go [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+// Combine joins the executed suffix with the cached prefix into a new [BlockTrace].
+//
+// Parameters:
+//   - executed: the VM result packaged by [FromVM] after executing
+//     transactions[Start:End]. It must contain one trace per transaction,
+//     in block order, with a non-nil state diff in each trace.
+//
+// The result is marked complete only if this range reaches the block's end.
+func (r *Range) Combine(executed *BlockTrace) (*BlockTrace, error) {
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:124]: I believe this method should have a `*BlockTrace` receiver rather than a `*Range`. 

For me, it makes more sense that a `BlockTrace` uses a new range to create an expanded copy of itself, and the constructor is defined in the same type. 

What do you think?

File: rpc/tracecache/progressive.go [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+		if executed.Traces[index].vmTrace == nil || executed.Traces[index].vmTrace.StateDiff == nil {
+			return nil, fmt.Errorf("VM omitted state diff for transaction trace %d", r.Start+uint64(index))
+		}
+	}
+	result := *executed
+	result.Traces = make([]TransactionTrace, len(r.prefix)+len(executed.Traces))
+	copy(result.Traces, r.prefix)
+	copy(result.Traces[len(r.prefix):], executed.Traces)
+	result.Complete = r.End == r.total
+	return &result, nil
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:147]: Is copying InitialRead as they are, safe?

I was under the impression, which it might be wrong (please let me know if I am), that initial reads could update themselves by executing the rest of the transactions. 

I need to read more of the PR but I am a bit confused how initial reads interacts with this

File: rpc/tracecache/trace.go
Comment (rodrodros) [inline rpc/tracecache/trace.go:1]: A bit late on the PR stack, but there should be a `trace_test.go`, unit testing `BlockTrace` and `TransactionTrace` public API

It can be done in a follow up

File: rpc/tracecache/progressive.go (hunk 9 above)
Comment (rodrodros) [reply rpc/tracecache/progressive.go]: Also, it seems that `executed` is then being overwritten by the `result` (on the places this function is called on the follow up). 

I believe this accentuates the point that there is no reason for `Combine` to be on `Range` but as part of `BlockTrace` instead. Additionally, it begs the question if need to return a new `BlockTrace` instead of modifying the existing one? 

Maybe there is a reason for this that I am missing, kindly let me know if that's the case

File: rpc/tracecache/progressive.go [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+func (r *Range) ResumeState(
+	parent core.StateReader,
+	classes core.StateReader,
+	blockNumber uint64,
+) (core.StateReader, error) {
+	if r.Start == 0 {
+		return parent, nil
+	}
+	checkpoint := checkpointFromTraces(r.prefix)
+	declared, err := loadCheckpointClasses(&checkpoint, classes)
```
Comment (rodrodros) [inline rpc/tracecache/progressive.go:109]: Do you think it would make sense to cache the return values of `checkpointFromTraces` and `loadCheckpointClasses` results to avoid re-doing the prefix each time?

Maybe it doesn't make sense. Just asking if this was considered

File: rpc/tracecache/trace.go (hunk 6 above)
Comment (danielntmd) [reply rpc/tracecache/trace.go:64]: This makes sense, `CoversTarget` is now a nil pointer error in order differentiate the two and downstream consumers use each respective call. Unit tests added.

File: rpc/tracecache/trace.go (hunk 7 above)
Comment (danielntmd) [reply rpc/tracecache/trace.go:73]: Changed to `ValidTarget` and returns bool.

File: rpc/tracecache/progressive.go (hunk 9 above)
Comment (danielntmd) [reply rpc/tracecache/progressive.go]: No particular reason for keeping the new executed result to be immutable. Then we can change the method to have the `*BlockTrace`  receiver. 

And I agree we can avoid the shallow struct copy by reusing the existing one, traces will remain copied for reader access.

File: rpc/tracecache/progressive.go (hunk 10 above)
Comment (danielntmd) [reply rpc/tracecache/progressive.go]: Yes, safe under current execution flow. 

1. `starknet_traceTransaction` [itself](https://raw.githubusercontent.com/starkware-libs/starknet-specs/master/api/starknet_trace_api_openrpc.json) does not accept an initial reads flag and always passes false internally. Only path for initial reads is through `starknet_traceBlockTransactions' which will always replay the full block.

2. This guard will skip setting the prefix and will leave `prefix` nil and `Start` at zero.
`rpc/tracecache/progressive.go:74`
`  if cached != nil && !initialReads {
      // ....
      plan.prefix = cached.Traces
      plan.Start = uint64(len(plan.prefix))
  }`
  
The behavior of replaying on initial reads is the same as before and documented in `PlanRange` `rpc/tracecache/progressive.go:52`.

File: rpc/tracecache/trace.go
Comment (danielntmd) [reply rpc/tracecache/trace.go:1]: Added into this PR, just to keep everything consolidated.

File: rpc/tracecache/progressive.go (hunk 11 above)
Comment (danielntmd) [reply rpc/tracecache/progressive.go:109]: I started with caching that state diff, but then we have essentially duplicated data cached. 

The tradeoff of marginal runtime execution overhead versus memory growth favors the latter. 

This is only for extending the cache, cache hits don't need to run through this. 

For `loadCheckpointClasses`:
  | Class sample | Extra memory retained by caching | Load time avoided | 
  |---|---:|---:|
  | Small | 10.91 KiB | 16.32 µs | 
  | Medium | 144.91 KiB | 183.11 µs | 
  | Large | 2.76 MiB | 3.33 ms | 
  
For `checkpointFromTraces`:
Microbenchmark on mainnet blocks: 
| Prefix | Unique slots | Baseline retained | Extra cached | Rebuild time avoided |
  |---|---:|---:|---:|---:|
  | Typical: 3 tx | 17 | 3.20 KiB | +3.62 KiB | 1.42 µs | 
  | p90-sized: 16 tx | 118 | 21.34 KiB | +17.74 KiB | 10.37 µs | 
  | Largest sampled: 2 tx | 2,116 | 177.81 KiB | +261.39 KiB | 76.35 µs | 
  
 e2e k6 (Basically noise): 
    | Workload | Recompute latency | Cached latency | Throughput change | CPU change |
  |---|---:|---:|---:|---:|
  | Transaction only | 544.7 ms | 543.0 ms | +0.26% | -0.19% |
  | Transaction → block | 1,092.6 ms | 1,091.8 ms | +0.09% | -0.16% |
  | Beginning → middle → end | 1,095.3 ms | 1,090.9 ms | +0.47% | -0.25% |

Comment (rodrodros) [reply rpc/tracecache/progressive.go:109]: We can keep them as they are then 👍

**Review by rodrodros — APPROVED** [review]
